include/boost/url/bnf/parse.hpp

```diff
@@ -16,58 +16,77 @@
 #include <boost/type_traits/make_void.hpp>
 #include <type_traits>
 
+/** @file
+
+   Parsing commands
+
+   This file contains the basic overloads for parsing BNF elements
+
+   @par Design of BNF elements
+
+   A BNF element implements the rules for parsing a BNF grammar rule and stores parsing results.
+
+   A BNF element:
+
+    - Is default constructible
+    - Holds the result, usually as a reference, of parsing
+    - Has a nested value_type
+    - Has `value_type& operator*() const noexcept`
+
+   Users should define a free function `parse` as a customization point defining how to parse their
+   BNF elements as part of the same architecture that might include arbitrary BNF elements in expressions.
+
+   These new function overloads may be defined in other namespaces. As with `std::swap`, the design relies
+   on Argument Dependent Lookup (ADL) to find these overloads.
+
+*/
+
+
 namespace boost {
 namespace urls {
 namespace bnf {
 
-/*  Design
-
-    - An element is default constructible
-    - Holds the result of parsing
-    - overloads the free function `parse`
-    - has a nested value_type
-    - has `value_type& operator*() const noexcept`
-*/
 
-/** @brief Parse a literal character
+/** Parse beginning of a char range a BNF literal character
 
    This function parses the string defined by the first chars in the
-   range of chars [@ref it, @ref end) as a literal character.
+   range of chars [`it`, `end`) as a literal character.
 
-   The function returns `true` and resets @ref ec when the input is
-   parsed successfully. The initial character in the range is immediately
-   consumed and ignored.
-
-   If the input cannot be parsed into the literal BNF literal,
-   the function returns `false` and sets the error number in @ref ec to the
-   appropriate error number defined in @ref boost::urls::error.
+   The initial character in the range is immediately consumed and ignored.
 
    @par Example
+
    @code
    std::string str = "Hello.cpp";
    auto it = str.data();
    auto const end = it + str.size();
    if (bnf::parse(it, end, ec, 'H')) {
-       std::cout << "str begins with 'H'" << std::endl;
+       std::cout << "str begins with 'H'\n";
    }
    @endcode
 
    @par Exception Safety
-   Calls to other `parse` overloads may throw, although they should
-   prefer setting @ref ec instead.
 
-   @return `true` if the first char in the range matches @ref ch successfully,
+   Never throws
+
+   @see
+
+   [Backus–Naur form](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form),
+
+   @ref boost::urls::error
```

Comment:
Prefer `@ref error`. There's no need to write out the namespace. However, note that there are two enumerations in this library. One in `boost::urls::error` and the other in `boost::urls::bnf::error`. Because `urls::bnf` is logically a small, separate "mini-library." That is, a dependent library can use `bnf` algorithms without using the `urls` algorithms.